core/trino-main/src/main/java/io/trino/operator/join/BigintPagesHash.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.operator.join;
+
+import com.google.common.collect.ImmutableList;
+import io.airlift.units.DataSize;
+import io.trino.operator.HashArraySizeSupplier;
+import io.trino.operator.PagesHashStrategy;
+import io.trino.spi.Page;
+import io.trino.spi.PageBuilder;
+import io.trino.spi.block.Block;
+import it.unimi.dsi.fastutil.longs.LongArrayList;
+import org.openjdk.jol.info.ClassLayout;
+
+import java.util.Arrays;
+import java.util.List;
+
+import static io.airlift.slice.SizeOf.sizeOf;
+import static io.airlift.units.DataSize.Unit.KILOBYTE;
+import static io.trino.operator.SyntheticAddress.decodePosition;
+import static io.trino.operator.SyntheticAddress.decodeSliceIndex;
+import static io.trino.util.HashCollisionsEstimator.estimateNumberOfHashCollisions;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+
+// This implementation assumes:
+// -There is only one join channel and it is of type bigint
+// -arrays used in the hash are always a power of 2.
+public final class BigintPagesHash
+        implements PagesHash
+{
+    private static final int INSTANCE_SIZE = ClassLayout.parseClass(BigintPagesHash.class).instanceSize();
+    private static final DataSize CACHE_SIZE = DataSize.of(128, KILOBYTE);
+
+    private final LongArrayList addresses;
+    private final List<Block> joinChannelBlocks;
+    private final PagesHashStrategy pagesHashStrategy;
+
+    private final int mask;
+    private final int[] key;
+    private final long[] values;
```

Comment:
this array is redundant to the one in `pagesHashStrategy`. I think we can get rid of the array in `pagesHashStrategy`.

Reply:
The `PagesHash` is still used to produce the output page in `appenTo` method so we can get rid of it only if it does not exist in the output. Working on that. I hope that this can be merged as it is and then we can work on further reducing the memory footprint.

Reply:
> I hope that this can be merged as it is and then we can work on further reducing the memory footprint.

I think we can merge it only if we use bigint path on small hash tables. Otherwise there will still be increased memory usage.